src/core/dataframe.js

```diff
@@ -189,22 +189,15 @@ export default class Dataframe {
         // The viewport is in the [-1,1] range (on Y axis), therefore a pixel is equal to the range size (2) divided by the viewport height in pixels
```

Comment:
Hi :) I think we can use this PR to do a quick refactor here. `computeScale` method is duplicated (the only difference is that for lines it uses `vizStrokeWidth` and for polygons `vizWidth`). Also, `_getLinesAtPosition` and `_getPolygonsAtPosition` are also similar, so I propose the following refactor:

1. Add a function to get the scale value instead of modifying it:

```js
    _getScale(lineWidth, featureIndex, columnNames, widthScale) {
        const LINE_SATURATION_PIXELS = 336;
        const features = {};
        
        columnNames.forEach(name => {
            features[name] = this.properties[name][featureIndex];
        });

        const widthDiameter = Math.min(lineWidth.eval(features), LINE_SATURATION_PIXELS);
        const widthRadius = widthDiameter / 2;
        
        return widthRadius * widthScale;
    }
```

2. Change `_addFeatureToArray` to `_getFeatureProperties` and return only the feature properties, in order to avoid modifying the array in this function.

```js
    _getFeatureProperties(featureIndex) {
        // same code
        
        return { id, properties };
    }
```

3. Add a function to get the geometry values that are being used in `_getPointsAtPosition`, `_getLinesAtPosition` and `_getPolygonsAtPosition`.

```js
    _getGeometryValues(position) {
        const pos = wToR(position.x, position.y, { center: this.center, scale: this.scale });
        const points = this.decodedGeom.vertices;

        // The viewport is in the [-1,1] range (on Y axis), therefore a pixel is equal to the range size (2) divided by the viewport height in pixels
        const widthScale = (2 / this.renderer.gl.canvas.clientHeight) / this.scale * this.renderer._zoom;
        const columnNames = Object.keys(this.properties);

        return { pos, points, widthScale, columnNames };
    }
```

3. Add a function that shares the behaviour of `_getLinesAtPosition` and `_getPolygonsAtPosition` to avoid having the same code duplicated

```js
    _getLinesAtPosition(position, viz) {
        const lineWidth = viz.width;

        return this._getGeometryAtPosition(position, lineWidth);
    }

    _getPolygonAtPosition(position, viz) {
        const lineWidth = viz.strokeWidth;

        return this._getGeometryAtPosition(position, lineWidth);
    }
```

Where `_getGeometryAtPosition` will include all the changes above:

```js
    _getGeometryAtPosition(position, lineWidth) {
        // Linear search for all features
        // Tests triangles instead of polygons since we already have the triangulated form
        // Moreover, with an acceleration structure and triangle testing features can be subdivided easily
        const features = [];
        const vertices = this.decodedGeom.vertices;
        const normals = this.decodedGeom.normals;
        const breakpoints = this.decodedGeom.breakpoints;
        const { pos, widthScale, columnNames } = this._getGeometryValues(position);
        
        let featureIndex = 0;
        let scale = this._getScale(lineWidth, featureIndex, columnNames, widthScale);

        const TRIANGLES_EDGES = 6;

        for (let i = 0; i < vertices.length; i += TRIANGLES_EDGES) {
            
            if (i >= breakpoints[featureIndex]) {
                featureIndex++;
                scale = this._getScale(lineWidth, featureIndex, columnNames, widthScale);
            }

            const v1 = {
                x: vertices[i + 0] + normals[i + 0] * scale,
                y: vertices[i + 1] + normals[i + 1] * scale
            };
            const v2 = {
                x: vertices[i + 2] + normals[i + 2] * scale,
                y: vertices[i + 3] + normals[i + 3] * scale
            };
            const v3 = {
                x: vertices[i + 4] + normals[i + 4] * scale,
                y: vertices[i + 5] + normals[i + 5] * scale
            };

            const inside = pointInTriangle(pos, v1, v2, v3);
            
            if (inside) {
                const feature = this._getFeatureProperties(featureIndex);
                features.push(feature);
                // Don't repeat a feature if we the point is on a shared (by two triangles) edge
                // Also, don't waste CPU cycles
                i = breakpoints[featureIndex] - TRIANGLES_EDGES;
            }
        }

        return features;
    }
```

What do you think? 😄

Reply:
I already addressed number 3 in https://github.com/CartoDB/carto-vl/pull/539 . I addressed number 1 in this PR as requested by @IagoLast . But on the #539 PR I went back to the inline function, since the refactor number 3 removes that duplication too, and it can be confusing, since that scale is not applicable for points, therefore, it's better to be with the code for polygons and lines only. 

Regarding number 2, I agree, but, in any case, I would do it on #539

Reply:
👍